src/peft/tuners/lora/model.py

```diff
@@ -164,7 +164,8 @@ def _prepare_model(self, peft_config: LoraConfig, model: nn.Module):
             model (`nn.Module`):
                 The model that is going to be adapted.
         """
-        if peft_config.layer_replication:
+        # Handle X-LoRA case
```

Comment:
No, those cannot because during init those functions are called.

Reply:
I think I found a way to allow us to remove all these changes to `lora/model.py`. The main issue is the following: During `XLoraModel.__init__`, we want to create the `LoraModel` instance based on the `XLoraConfig`. This `LoraModel` is not supposed to contain any actual LoRA adapter, as those will come later through the `config.adapters`. So what we need is to implement a way to create this "empty" `LoraModel` without needing to change `LoraModel` itself. Here is my idea:

First, let's make a change to `XLoraModel.__init__` by creating a copy of the `XLoraConfig` that "imitates" a normal `LoraConfig`:

```diff
modified   src/peft/tuners/xlora/model.py
@@ -140,7 +140,15 @@ class XLoraModel(BaseTuner):
             conf = config[adapter_name]
         else:
             conf = config
-        lora_model = LoraModel(model, config.copy(), adapter_name)
+
+        # create an empty LoraModel
+        base_lora_config = copy.copy(conf)
+        base_lora_config.target_modules = DUMMY_TARGET_MODULES
+        # imitate a LoraConfig, fields might need to be updated if LoraConfig is updated
+        base_lora_config.layer_replication = None
+        base_lora_config.bias = "none"
+        lora_model = LoraModel(model, base_lora_config, adapter_name)
+
```

So we set the required attributes so that we no longer need the extra checks in `LoraModel` that this PR adds. Also, we have added `DUMMY_TARGET_MODULES`. What is this? It's a constant defined in `constants.py` as `DUMMY_TARGET_MODULES = "dummy-target-modules"`. This is a special value that allows us to create an "empty" `LoraModel`. For this to work, we also need a small change in `BaseTuner`:

```diff
modified   src/peft/tuners/tuners_utils.py
@@ -395,6 +395,10 @@ class BaseTuner(nn.Module, ABC):
         self._prepare_model(peft_config, model)
         is_target_modules_in_base_model = False
         key_list = [key for key, _ in model.named_modules()]
+        if getattr(peft_config, "target_modules", None) == DUMMY_TARGET_MODULES:
+            # dummy adapter, we allow not matching any module
+            key_list = []
+            is_target_modules_in_base_model = True
```

I tested this locally and the tests pass after reverting all changes to `lora/model.py`.

I think this is the more elegant solution, because it should not be necessary for `LoraModel` to know about XLoRA.

Reply:
That sounds good, I'll add those 2 changes - this means I can revert all changes to lora/model.py to `main`, right?

Reply:
Exactly, that's the idea. You'd also have to add `DUMMY_TARGET_MODULES` to `utils/constants.py` and add a few imports, the rest should work.

Reply:
Great! I pushed those changes now, and all the tests pass locally.